### DISCUSSION TITLE
Inlining `Vec`, `HashMap`, `Option` and `Result`

### DISCUSSION THREAD

**Discussion by escritorio-gustavo** [Aleph-Alpha/ts-rs#239] [General]
The four types in the title (as well as anything that uses `impl_shadow!` to defer to them) cause the `#[ts(inlne)]` attribute to generate invalid import statements, as shown in #168 and #232. This discussion is meant to decide how to tackle this problem.

**Comment by escritorio-gustavo** [comment]
The following is one of the ideas I proposed in #233:

For `Vec` and `HashMap`, we could make `Self::inline` just return `Self::name` or `panic!` saying they can't be `inline`d

This idea seems ridiculous, but think about it:
- `Vec<T>::name` returns `Array<T::name()>`
- `Vec<T>::inline` returns `Array<T::inline()>`

But notice that `Vec<T>::inline` doesn't actually inline `Vec`, instead it inlines `T`.
Why? Because there is no way to inline `Array`.

The problem is instead of stating this to be illegal (or just the same as calling name), we recursively
inline `T`, but our derive macro can't handle that!

The same is true for `Option` and `Result`. `Self::name` returns them already inlined! With any other enum, inlining gives the type union representation, while these two do that for name. There is no inlining an enum's type union form!

In fact, if Option were exported, name would return `Option<T::name()>` and inline would return what name currently returns

You can even test this out by creating your own Option and Result enums:
```rs
#[derive(TS)]
#[ts(export, untagged)]
enum MyOption<T> {
    Some(T),
    None
}

#[derive(TS)]
#[ts(export, rename_all = "lowercase")]
enum MyResult<T, E> {
    Ok(T),
    Err(E),
}

#[derive(TS)]
#[ts(export)]
struct MyStruct {
    not_inlined_option: MyOption<String>,

    #[ts(inline)]
    inlined_option: MyOption<String>,

    not_inlined_result: MyResult<EnumWithName, String>,

    #[ts(inline)]
    inlined_result: MyResult<EnumWithName, String>,
}

#[test]
fn check() {
    assert_eq!(
        MyStruct::decl(),
        "type MyStruct = { not_inlined_option: MyOption<string>, inlined_option: string | null, not_inlined_result: MyResult<EnumWithName, string>, inlined_result: { \"ok\": EnumWithName } | { \"err\": string }, };"
    )
}
```

**Reply by escritorio-gustavo** [reply]
And here is the exported TS, which contains all the correct `import`s

```ts
// This file was generated by [ts-rs](https://github.com/Aleph-Alpha/ts-rs). Do not edit this file manually.
import type { EnumWithName } from "./EnumWithName"; // from #168
import type { MyOption } from "./MyOption";
import type { MyResult } from "./MyResult";

export type MyStruct = {
  not_inlined_option: MyOption<string>;
  inlined_option: string | null;
  not_inlined_result: MyResult<EnumWithName, string>;
  inlined_result: { "ok": EnumWithName } | { "err": string };
};
```

**Comment by NyxCode** [comment]
I think your analysis here is correct to some extent. The handeling for `Option` and `Result` is different such that they're always being inlined.

To make them behave like every other type, we could just define `#[derive(TS)] enum ResultImpl { ... }`, and then use `impl_shadow!(as ResultImpl<T, E>: impl<T: TS, E: TS> TS for Result<T, V>);` to make `Result<T, E>` behave just like our `ResultImpl`.  

However, it seems like there's an other bug lurking here:
```rs
#[derive(TS)]
#[ts(export)]
struct Test {}

#[derive(TS)]
#[ts(export)]
enum MyOption<T> {
    Some(T),
    None
}

#[derive(TS)]
#[ts(export)]
struct Outer {
    a: MyOption<Test>
}
```
generates this TS, in which the import for `Test` is missing!
```ts
// This file was generated by [ts-rs](https://github.com/Aleph-Alpha/ts-rs). Do not edit this file manually.
import type { MyOption } from "./MyOption";

export type Outer = { a: MyOption<Test>, };
```

I'm not 100% sure about this, but fixing this bug might allow us to keep our current special-handeling of `Option` and `Result`.

**Reply by NyxCode** [reply]
The dependencies in this example are as follows:  
`MyOption<T>`: `[T]`
`Outer:`: `[MyOption<T>]`

Since `MyOption` is not transparent, it's dependencies aren't passed through, so exporting `Outer` only causes `MyOption` to be imported.  

It's not immediately clear to me what the most elegant fix for this would be.   
One solution might be to introduce a `name_dependency_types` function, but that seems a bit ugly. 
Alternatively, when deriving `TS` for `Outer`, we could try to add the generic of `MyOption` (`T`) to the dependencies directly, making 
`Outer::dependency_types` return `[MyOption<Test>, Test]`.

**Reply by escritorio-gustavo** [reply]
We could also have a `#[ts(trasparent)]` attribute to make `MyOption` transparent

**Reply by escritorio-gustavo** [reply]
Hold up. Doesn't the problem you described happen for every generic type then?

**Reply by NyxCode** [reply]
> Hold up. Doesn't the problem you described happen for every generic type then?

Seems so - if the generic parameter of that type is a primitive, then everything's fine, but if it would need to be imported, it breaks.  

Just checked, this was already broken before the last PR. I'll create a branch with a few tests, so we can figure out how to fix it.

**Reply by NyxCode** [reply]
In #241, I reproduced the bug in one of the end-to-end tests.  
In the example, the bug was already present, but we didn't run CI on it before.

**Reply by escritorio-gustavo** [reply]
I managed to get this fixed  (Example you showed now correct) by adding this to `macros/src/types/named.rs`
```rs
            if let Type::Path(T { path, .. }) = ty {
                path
                    .segments
                    .last()
                    .map(|x| &x.arguments)
                    .and_then(|x| if let PathArguments::AngleBracketed(A { ref args, ..}) = x {
                        Some(
                            args
                                .iter()
                                .filter_map(|x| if let GenericArgument::Type(t) = x {
                                    Some(t)
                                } else {
                                    None
                                })
                                .for_each(|x| dependencies.push(x))
                        )
                    } else {
                        None
                    });
            }
```

**Reply by escritorio-gustavo** [reply]
My guess is this needs to be done everywhere that `inline` affects dependencies

**Reply by NyxCode** [reply]
That was my first idea as well, nice!
I think I outlined the problem with that approach in #241 - type aliases.

**Reply by escritorio-gustavo** [reply]
> I think I outlined the problem with that approach in https://github.com/Aleph-Alpha/ts-rs/pull/241 - type aliases.

Makes sense, and then we'd be right back to #70

**Reply by NyxCode** [reply]
Only for imports though, so I think we are making progress :D

**Reply by escritorio-gustavo** [reply]
Okay, so we need to make sure to add the `generics()` to the import, but only when the type is not inlined, right?

**Reply by escritorio-gustavo** [reply]
I imagine this can be done by changing `Dependencies::push` to also add `T::generics()`?

**Reply by NyxCode** [reply]
Exactly! That's why I put that logic in `Dependencies::push`. If a type is inlined, we don't use `Dependencies::push`, since the parent type doesn't actually depend on the type that got inlined. In that case, we call `Dependencies::append_from`.

**Comment by escritorio-gustavo** [comment]
Alright, now that #241 has resolved the problem with generics not being imported, we need to get back to how we are supposed to deal with the inlining of transparent types (`Option`, `Result`, `Vec`, `HashMap` and their `impl_shadows`)

**Comment by NyxCode** [comment]
Absolutely!

After thinking about your analysis a bit, I now totally agree that the recursive inlining the impls currently do is a no-go since it breaks the dependency tracking outright. 

Let's look at `Result<T, E>`:

It currently returns `format!("{{ Ok : {} }} | {{ Err : {} }}", T::name(), E::name())` as `TS::name()`. That's not too different from `"Result<{T::name()}, {E::name()}>"`, which is what a `#[derive(TS)] enum Result<T, E> { .. }` type would return. In both cases, the `TS::name()`s of `T` and `E` end up in the `TS::name()` of `Result<T, E>`.  
If inlining here was just a no-op, I think that impl would be fine.

**Reply by escritorio-gustavo** [reply]
Yeah, there's no way to track the dependencies if we ever do `T::inline()` and `E::inline()`.

My idea is to make **both** `Result::<T, E>::name()` and `Result::<T, E>::inline()` return `format!("{{ Ok : {} }} | {{ Err : {} }}", T::name(), E::name())`

**Comment by NyxCode** [comment]
Also, out of the four (Vec, HashMap, Option, Result), Result is different than the other 3.  
The others are primitives in TS, and we actually need a custom `TS` impl for them. 

`Result`, on the other hand, we *could* get with `#[derive(TS)]`

**Reply by escritorio-gustavo** [reply]
I think your idea of having `Result` be an `impl_shadow` of a derived `ResultDef` is the best way to go

**Comment by escritorio-gustavo** [comment]
@NyxCode, I have pushed a commit that should fix the problem by making all 4 of those impls more closely resemble `impl_wrapper` over their generics

**Comment by escritorio-gustavo** [comment]
This has been successfully resolved by merging #241
